librz/util/buf_io.c

```diff
@@ -1,82 +1,58 @@
-// SPDX-FileCopyrightText: 2009-2020 ret2libc <sirmy15@gmail.com>
+// SPDX-FileCopyrightText: 2021 Florian Märkl <info@florianmaerkl.de>
 // SPDX-License-Identifier: LGPL-3.0-only
 
 #include <rz_util.h>
 #include <rz_io.h>
 
-struct buf_io_user {
-	RzIOBind *iob;
-	int fd;
-};
+typedef RzIOBind *BufIOUser;
 
-struct buf_io_priv {
+typedef struct buf_io_priv {
 	RzIOBind *iob;
-	int fd;
-};
-
-static inline struct buf_io_priv *get_priv_io(RzBuffer *b) {
-	struct buf_io_priv *priv = (struct buf_io_priv *)b->priv;
-	rz_warn_if_fail(priv);
-	return priv;
-}
+	ut64 offset;
+} BufIOPriv;
 
 static bool buf_io_init(RzBuffer *b, const void *user) {
-	const struct buf_io_user *u = (const struct buf_io_user *)user;
-	struct buf_io_priv *priv = RZ_NEW0(struct buf_io_priv);
+	BufIOUser u = (void *)user;
+	rz_return_val_if_fail(u, false);
+	BufIOPriv *priv = RZ_NEW(BufIOPriv);
 	if (!priv) {
 		return false;
 	}
-	priv->iob = u->iob;
-	priv->fd = u->fd;
+	priv->iob = u;
+	priv->offset = 0;
 	b->priv = priv;
-	b->fd = priv->fd;
 	return true;
 }
 
 static bool buf_io_fini(RzBuffer *b) {
-	// struct buf_io_priv *priv = get_priv_io (b);
-	RZ_FREE(b->priv);
+	free(b->priv);
 	return true;
 }
 
 static st64 buf_io_seek(RzBuffer *b, st64 addr, int whence) {
-	struct buf_io_priv *priv = get_priv_io(b);
-	int io_whence;
-
-	switch (whence) {
-	default:
-		rz_warn_if_reached();
-	case RZ_BUF_SET:
-		io_whence = RZ_IO_SEEK_SET;
-		break;
-	case RZ_BUF_END:
-		io_whence = RZ_IO_SEEK_END;
-		break;
-	case RZ_BUF_CUR:
-		io_whence = RZ_IO_SEEK_CUR;
-		break;
-	}
-	return priv->iob->fd_seek(priv->iob->io, priv->fd, addr, io_whence);
+	BufIOPriv *priv = b->priv;
+	priv->offset = rz_seek_offset(priv->offset, 0, addr, whence);
+	return priv->offset; // TODO: this might be negative and be detected as error
 }
 
 static ut64 buf_io_get_size(RzBuffer *b) {
-	struct buf_io_priv *priv = get_priv_io(b);
-	return priv->iob->fd_size(priv->iob->io, priv->fd);
+	return 0; // TODO: overflow :-(
```

Comment:
?

Reply:
io has a size of 2^64, which we can't express. I've just removed the size callback entirely since not implementing it is supported by RzBuffer too.